### ISSUE TITLE
blockchain/v0: possible data race/panic

### ISSUE THREAD

**Issue by cuonglm** [tendermint/tendermint#6516]
**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source):

1e0efb675899a9217d289db02c3f226b54e036c4

**What happened**:

Data race in Github Action with following command:

```
cat pkgs.txt.part.00 | xargs go test -mod=readonly -timeout 8m -race -coverprofile=00profile.out -covermode=atomic
```

https://github.com/tendermint/tendermint/pull/6509/checks?check_run_id=2708736360

**What you expected to happen**:

Test passed.

**Have you tried the latest version**: yes

**How to reproduce it** (as minimally and precisely as possible):

It's hard to reproduce this consistently in local. But I can see possible problem through code review.

`r.processBlockchainCh`:

https://github.com/tendermint/tendermint/blob/7bf84d9d7f6f80a153e29cb4f1fd17b0d4264f1a/blockchain/v0/reactor.go#L159

and `r.processPeerUpdates`:

https://github.com/tendermint/tendermint/blob/7bf84d9d7f6f80a153e29cb4f1fd17b0d4264f1a/blockchain/v0/reactor.go#L160

are run in two separated goroutines. `r.processPeerUpdates` will call `r.processPeerUpdate`, which will send to `r.blockchainCh.Out`:

https://github.com/tendermint/tendermint/blob/7bf84d9d7f6f80a153e29cb4f1fd17b0d4264f1a/blockchain/v0/reactor.go#L317

While `r.processBlockchainCh` will close that channel:

https://github.com/tendermint/tendermint/blob/7bf84d9d7f6f80a153e29cb4f1fd17b0d4264f1a/blockchain/v0/reactor.go#L285

In summary, we send to a channel in a goroutine but close it in another one. That's exactly what the error in Github Action reports.

In the worst case, that would lead to a panic.

**Comment by cuonglm** [comment]
cc @odeke-em

**Comment by cmwaters** [comment]
I can confirm that I've seen this race as well

**Comment by cuonglm** [comment]
@alexanderbez @cmwaters I realize the same problem happens with `requestRoutine` and `processBlockchainCh`. If the approach in #6518 were approved, we can do the same trick for this.

**Comment by marbar3778** [comment]
@cuonglm can this be closed?

**Comment by cuonglm** [comment]
> @cuonglm can this be closed?
> 
> 

@marbar3778 not yet, there's same issue in `requestRoutine` need to be fixed.
